### ISSUE TITLE
VerificationError - bitcoin verification failed: Error: ETIMEDOUT

### ISSUE THREAD

**Issue by ethicnology** [opentimestamps/javascript-opentimestamps#73]
Hi there,  
I'm working on opentimestamps and i observe a problem when using the verify function.
I set ignore bitcoin node option so it use default blockstream esplora instead.
The result **is unstable**, sometimes i have the expected return :
```
        "bitcoin": {
            "timestamp": 1591975254,
            "height": 634392
         }
```
more often it return this error :
```
Lite-client verification, assuming block 00000000000000000006e893f9299d44c8b0e66b5ecd9576c20a38f326bcd7be is valid
Lite-client verification, assuming block 0000000000000000000df209a52dad91f09a43b0254db81f4e00d79f56827b45 is valid
Lite-client verification, assuming block 0000000000000000000d0f41d23b1475fd5ee4e3fe9455ae1fe6bc8b131b44be is valid
Lite-client verification, assuming block 00000000000000000010bf4a92e227aef948f0ee822a993fdcefd67d4a1fc4e5 is valid
Lite-client verification, assuming block 0000000000000000000df209a52dad91f09a43b0254db81f4e00d79f56827b45 is valid
Lite-client verification, assuming block 00000000000000000006e893f9299d44c8b0e66b5ecd9576c20a38f326bcd7be is valid
Lite-client verification, assuming block 00000000000000000010bf4a92e227aef948f0ee822a993fdcefd67d4a1fc4e5 is valid
Response error: RequestError: Error: ETIMEDOUT
Response error: RequestError: Error: ETIMEDOUT
Response error: RequestError: Error: ETIMEDOUT
Response error: RequestError: Error: ETIMEDOUT
VerificationVerificationError: 'bitcoin verification failed: Error: ETIMEDOUT': bitcoin verification failed: Error: ETIMEDOUT
```
Running a local node is it the only way to bypass this error?

**Comment by lvaccaro** [comment]
Hi @ethicnology, sorry to be late.
You could increase the request timeout time by passing a timeout value in options to `verify()` function as:
```js
let options = {};
options.ignoreBitcoinNode = true;
options.timeout = 5000;
OpenTimestamps.verify(detachedOts,detached,options)
```
Moreover it should be good expose timeout as a parameter from `ots-cli.js` script.

Moreover, I don't know if this is the case, but esplora should have a limit threshold after a number of requests in short time.

**Comment by ethicnology** [comment]
Thank @lvaccaro for the answer, unfortunately i think there is another issue here.
I spent a lot of time to review [open-timestamp.js](https://github.com/opentimestamps/javascript-opentimestamps/blob/master/src/open-timestamps.js) code file to make PR #72 and i had no memories that there is a **timeout** parameter.
I double checked and there is no **options.timeout** documented/used in this file howeover, i found **this.timeout** parameter used in the [esplora.js constructor](https://github.com/opentimestamps/javascript-opentimestamps/blob/efaaae1cfd31fd362fe4951822eb97929a623343/src/esplora.js#L27) which is called during the verify process, [here](https://github.com/opentimestamps/javascript-opentimestamps/blob/efaaae1cfd31fd362fe4951822eb97929a623343/src/open-timestamps.js#L318).
The fact is that the constructor receive an [empty options parameter](https://github.com/opentimestamps/javascript-opentimestamps/blob/efaaae1cfd31fd362fe4951822eb97929a623343/src/esplora.js#L25) maybe because of a small mistake, [here](https://github.com/opentimestamps/javascript-opentimestamps/blob/efaaae1cfd31fd362fe4951822eb97929a623343/src/open-timestamps.js#L314).

```js
// src/open-timestamp.js
verifyAttestation (attestation, msg, options = {}) {          //The given options is erased
    function liteVerify (options = {}) {                       //here by the new declaration in liteVerify function
        options.chain = options.chain ? options.chain : 'bitcoin'
        const esplora = new Esplora(options)
        ...
        ...
}
```

To verify my words, i cloned the project and i added a **console.log(options)** in [liteVerify](https://github.com/opentimestamps/javascript-opentimestamps/blob/efaaae1cfd31fd362fe4951822eb97929a623343/src/open-timestamps.js#L314) and in [esplora constructor ](https://github.com/opentimestamps/javascript-opentimestamps/blob/efaaae1cfd31fd362fe4951822eb97929a623343/src/esplora.js#L25): 
```js
 // src/open-timestamp.js
  verifyAttestation(attestation, msg, options = {}) {
    console.log("verifyAttestation : before", options)
    function liteVerify(options = {}) {
      console.log("verifyAttestation : after", options)
      ...
   }
```

```js
// src/esplora.js
  constructor(options = {}) {
    console.log("esplora : ", options)
    this.url = options.url ? options.url : PUBLIC_ESPLORA_URL
    this.timeout = options.timeout ? options.timeout : 1000
  }
```
 
Then i run a verify with the -i ignore-bitcoin-node argument to see if the **options.ignoreBitcoinNode stay true** in the esplora constructor console.log : 
```sh
node ots-cli.js verify -i file.txt.ots
```
Options  - options are erased : 
```sh
verifyAttestation - before : options = Command {
  ... ,
  ignoreBitcoinNode: true,
  ... ,
}
verifyAttestation - after : options =  {}
esplora : options =  { chain: 'bitcoin' }
```

To correct this behavior, we should remove **options = {}** from the liteVerify function declaration in there : 
```js
// src/open-timestamp.js
function liteVerify () {
    ...
}
```

Output - options are not erased : 
```sh
verifyAttestation - before : options = Command {
  ... ,
  ignoreBitcoinNode: true,
  ... ,
}
verifyAttestation - after : options =  Command {
  ... ,
  ignoreBitcoinNode: true,
  ... ,
}
esplora : options =  Command {
  ... ,
  ignoreBitcoinNode: true,
  ... ,
}
```

**Comment by lvaccaro** [comment]
Thanks for all the review. Concept ack.

As you described, `liteVerify` takes the esplora sub-option instead all the options passed by the ots-cli.

https://github.com/opentimestamps/javascript-opentimestamps/blob/efaaae1cfd31fd362fe4951822eb97929a623343/src/open-timestamps.js#L353 

Passing the full options simplify the object itself (where timeout is set), even if ignoreBitcoinNode it is not used.

**Comment by ethicnology** [comment]
Thank you for advice and review.
Solved when #76 #77 will be merged
